### ISSUE TITLE
Webdav door closes connection when receiving "Expect: 100-Continue" request

### ISSUE THREAD

**Issue by bgounon** [dCache/dcache#5604]
Hi,

We are currently running mass data ingestion tests to our dCache pool using webdav. For this, we use the "Expect: 100-Continue" header in our PUT requests to ask for redirection location and avoid proxying all transfers through the webdav door. Multiple ingestion clients will run in parallel to achieve optimal performance.

We observe that the webdav door will close the HTTP connection every time it receives a request with this header (regardless of the method used). The issue is that a given client process will most likely re-request the webdav door after every upload, to obtain the next upload's location from dCache.

This means that when uploading X files to dCache, at least X connections will be open then closed between the client and the door, where a single one would actually suffice. We think it could impact upload performance when working with large amounts of files and many parallel ingestion tasks.

Is this expected behavior ? Can we configure dCache webdav door to not close connection after receiving a request with "Expect: 100-Continue" set ?

Thank you !

Example using cURL:
```
$ curl -L -I -X GET --capath /etc/grid-security/certificates -H "Content-Type: text/html" -H "Authorization: Bearer $MACAROON" -H "Expect: 100-Continue" https://ccdcacli236.in2p3.fr:2880/lsst/users/bgounon/tests/test001
HTTP/1.1 302 Found
Date: Fri, 30 Oct 2020 09:44:21 GMT
Server: dCache/5.2.21
Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Location: https://ccdcacli364.in2p3.fr:30724/pnfs/in2p3.fr/data/lsst/users/bgounon/tests/test001?dcache-http-uuid=110bc74c-792f-4a3c-9819-5019111d5b96
Connection: close

HTTP/1.1 200 OK
Accept-Ranges: bytes
Content-Length: 20
Content-Disposition: attachment;filename=test001
Content-Location: https://ccdcacli236.in2p3.fr:2880/lsst/users/bgounon/tests/test001
```

**Comment by paulmillar** [comment]
Just as an immediate observation, the curl command seems to be transferring only one file (and also, this is a GET rather than a PUT).

When the curl command terminates, all network connections are closed.  This is non-negotiable, it's part of how Linux/Unix works when running a command.

However, curl may be told to transfer multiple files in one invocation.  If you haven't already done so, you might like to investigate the curl options for transferring multiple files and see if that helps.

**Comment by bgounon** [comment]
Yes, I am aware of this. Sorry my example wasn't very clear.
It was just an illustration of the fact that we see `Connection: close` immediately after the `302` response is sent by the webdav door, and before curl actually terminates. I used GET here but the same behavior is observed with PUT.

We're not actually using curl for transfers, instead we use is an implementation based on Python requests where we manage how transfers are done.

**Comment by paulmillar** [comment]
OK, I've dug deep into the Jetty source code and now understand what is happening.

As you described, the problem is triggered by the client including the `Expect: 100-continue` request header and dCache responding with something other than the expected `100 Continue`  status code; in this case, the redirection (307) status code.

The behaviour you describe, closing the connection, is actually hard-coded into Jetty, and for good reasons.

An RFC-conforming client sending a request with the `Expect: 100-continue` request header is allowed to send the entity (in this case, the file's content) at any time, even before receiving `100 Continue` reply.  Indeed, RFC 2616 specifically encourages clients to do so if they don't receive the `100 Continue` response within a certain time limit.

The problem is that this introduces a race condition.  The client could start sending the entity (file's contents) before receiving the 307 redirection status code from dCache.  If dCache were to persist the connection then, in this scenario, it would be unable to distinguish between the client sending the file's content and the start of a new HTTP request.

As it happens, this problem was reported to Jetty in [bug report 446406](https://bugs.eclipse.org/bugs/show_bug.cgi?id=446406).  The jetty team closed the bug as INVALID, meaning Jetty must close the connection.

So, the unfortunate situation is that dCache cannot fix this problem: redirected PUT requests using Expect-100 must result in the connection being closed.

There is some hope that HTTP/2 will allow the server to persist connections under these circumstances.  So this would be something to check once dCache supports HTTP/2.  Unfortunately I currently don't have a date when this will be available.

**Comment by paulmillar** [comment]
Hi @bgounon ,

One further suggestion.  If you haven't already enabled it, I suggest you look into TLS session resumption and TLS session caching.

TLS session resumption and session caching allows a TLS client to avoid the time-consuming TLS handshake when reconnecting after a TCP disconnect.  Although this isn't a perfect solution, TLS session caching should allow the client to re-establish the connection with only a minor overhead above the TCP three-way handshake.  For making PUT requests that result in redirections, this may have a significant impact on performance.

TLS session resumption should be enabled on dCache.  My tests against prometheus suggest it does, indeed, work:
```
paul@sprocket:~$ echo | openssl s_client -connect prometheus.desy.de:443 -reconnect
[...]
drop connection and then reconnect
CONNECTED(00000003)
Verification: OK
---
Reused, TLSv1.2, Cipher is ECDHE-RSA-AES256-GCM-SHA384
[...]
---
DONE
paul@sprocket:~$ 
```

**Comment by bgounon** [comment]
Thank you @paulmillar, I will look into it
